src/PulseAudioManager.vala

```diff
@@ -42,50 +42,165 @@ public class Sound.PulseAudioManager : GLib.Object {
+        debug ("  %s", device.input? "input" : "output");
+        // #1 Set card profile
+        // Some sinks / sources are only available under certain card profiles,
+        // for example to switch between onboard speakers to hdmi
+        // the profile has to be switched from analog stereo to digital stereo.
+        // Attempt to find profiles that support both selected input and output
+        var other_device = device.input? default_output : default_input;
+        var profile_name = get_matching_profile (device, other_device);
+        // otherwise fall back to supporting this device only
+        if (profile_name == null) {
+            profile_name = device.profiles[0];
+        }
+        if (profile_name != device.card_active_profile_name) {
+            debug ("set card profile: %s > %s", device.card_active_profile_name, profile_name);
+            // switch profile to get sink for this device
+            yield set_card_profile_by_index (device.card_index, profile_name);
+            // wait for new card sink to appear
+            debug ("wait for card sink / source");
+            yield wait_for_update (device, device.input? "card-source-name" : "card-sink-name");
+        }
+        // #2 Set sink / source port
+        // Speakers and headphones can be different ports on the same sink
+        if (!device.input && device.port_name != device.card_sink_port_name) {
+            debug ("set sink port: %s > %s", device.card_sink_port_name, device.port_name);
+            // set sink port (enables switching between headphones and speakers for example)
+            yield set_sink_port_by_name (device.card_sink_name, device.port_name);
+        }
+        if (device.input && device.port_name != device.card_source_port_name) {
+            debug ("set source port: %s > %s", device.card_source_port_name, device.port_name);
+            yield set_source_port_by_name (device.card_source_name, device.port_name);
+        }
+        // #3 Wait for sink / source to appear for this device
+        if (!device.input && device.sink_name == null ||
+            device.input && device.source_name == null) {
+            debug ("wait for sink / source");
+            yield wait_for_update (device, device.input? "source-name" : "sink-name");
+        }
+        // #4 Set sink / source
+        // To for example switch between onboard speakers and bluetooth audio devices
+        if (!device.input && device.sink_name != default_sink_name) {
+            debug ("set sink: %s > %s", default_sink_name, device.sink_name);
+            yield set_default_sink (device.sink_name);
+        }
+        if (device.input && device.source_name != default_source_name) {
+            debug ("set source: %s > %s", default_source_name, device.source_name);
+            yield set_default_source (device.source_name);
+        }
+    }
+
+    private string? get_matching_profile (Device device1, Device device2) {
```

Comment:
Should be static and probably a part of the `Device` class.

Reply:
Alright, moved it to Device.